File: app/src/main/java/com/capitalnowapp/mobile/broadcastreceivers/SMSRetrieverBroadCastReceiver.java
package com.capitalnowapp.mobile.broadcastreceivers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.capitalnowapp.mobile.constants.Constants;
import com.capitalnowapp.mobile.interfaces.SMSListener;
import com.google.android.gms.auth.api.phone.SmsRetriever;
import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.common.api.Status;

public class SMSRetrieverBroadCastReceiver extends BroadcastReceiver {

    private static SMSListener smsListener;

    @Override
    public void onReceive(Context context, Intent intent) {
        try {
            if (SmsRetriever.SMS_RETRIEVED_ACTION.equals(intent.getAction())) {
                Bundle extras = intent.getExtras();
                Status status = (Status) extras.get(SmsRetriever.EXTRA_STATUS);
                try {
                    String message1 = (String) extras.get(SmsRetriever.EXTRA_SMS_MESSAGE);
                    Log.d("sms is...", "status is..." + status);
                    Log.d("sms is...", "msg is..." + message1);
                }catch (Exception e){
                    Log.d("sms is...", "exception is..." + e.getLocalizedMessage());
                }
                switch (status.getStatusCode()) {
                    case CommonStatusCodes.SUCCESS:
                        // Get SMS message contents
                        String message = (String) extras.get(SmsRetriever.EXTRA_SMS_MESSAGE);
                        // Extract one-time code from the message and complete verification
                        // by sending the code back to your server.

                        // Retrieve verification code from sms
                        String verificationCode = message.replace(Constants.SMS_RETRIEVER_HASH_KEY, "").replaceAll("[^0-9]", "");
                        if (smsListener != null)
                            smsListener.onOTPReceived(verificationCode);

                        break;
                    case CommonStatusCodes.TIMEOUT:
                        // Waiting for SMS timed out (5 minutes)
                        // Handle the error ...
                        if (smsListener != null)
                            smsListener.onOTPTimeOut();

                        break;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void bindListener(SMSListener listener) {
        smsListener = listener;
    }

    public static void unbindListener() {
        smsListener = null;
    }
}

File: app/src/main/java/com/capitalnowapp/mobile/kotlin/utils/SMSRetrieverBroadCastReceiver.kt
package com.capitalnowapp.mobile.kotlin.utils

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import com.capitalnowapp.mobile.interfaces.SMSListener
import com.google.android.gms.auth.api.phone.SmsRetriever
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.api.Status
import java.util.regex.Pattern

class SMSRetrieverBroadCastReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        Log.d(TAG, "onReceive: ")
        try {
            if (SmsRetriever.SMS_RETRIEVED_ACTION == intent.action) {
                val extras = intent.extras
                val mStatus = extras!![SmsRetriever.EXTRA_STATUS] as Status?
                when (mStatus!!.statusCode) {
                    CommonStatusCodes.SUCCESS -> {
                        // Get SMS message contents'
                        var message = extras[SmsRetriever.EXTRA_SMS_MESSAGE] as String?
                        message = extractDigits(message)
                        Log.d(TAG, "onReceive: success $message")
                        if (smsListener != null) smsListener!!.onOTPReceived(message)
                    }
                    CommonStatusCodes.TIMEOUT -> {
                        // Waiting for SMS timed out (5 minutes)
                        Log.d(TAG, "onReceive: failure")
                        if (smsListener != null) smsListener!!.onOTPTimeOut()
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    companion object {
        private const val TAG = "SmsBroadcastReceiver"
        private var smsListener: SMSListener? = null
        fun bindListener(listener: SMSListener?) {
            smsListener = listener
        }

        fun unbindListener() {
            smsListener = null
        }

        fun extractDigits(`in`: String?): String {
            val p = Pattern.compile("(\\d{6})")
            val m = p.matcher(`in`)
            return if (m.find()) {
                m.group(0)
            } else ""
        }
    }
}
